pkg/controller/registry/resolver/step_resolver.go

```diff
@@ -84,14 +86,83 @@ func NewOperatorStepResolver(lister operatorlister.OperatorLister, client versio
 	return stepResolver
 }
 
+// replacementChainEndsInFailure returns true if the last entry in a chain of operators is in the failed state.
+func replacementChainEndsInFailure(csv *v1alpha1.ClusterServiceVersion, csvToReplacement map[string]*v1alpha1.ClusterServiceVersion) bool {
+	if csv == nil {
+		return false
+	}
+	itr := csv
+	visited := map[string]struct{}{}
+	for {
+		// Visit the csv
+		visited[itr.GetName()] = struct{}{}
+
+		// Check for a replacement
+		next, ok := csvToReplacement[itr.GetName()]
+		if !ok {
+			break
+		}
+
+		// Check if we have visited the CSV before
+		if _, ok := visited[next.GetName()]; ok {
+			break
+		}
+
+		// Move iterator along replacement chain
+		itr = next
+	}
+	return itr.Status.Phase == v1alpha1.CSVPhaseFailed
+}
+
+func (r *OperatorStepResolver) CachePredicates(namespace string) ([]cache.Predicate, error) {
+	predicates := []cache.Predicate{}
+
+	operatorGroupList, err := r.client.OperatorsV1().OperatorGroups(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return predicates, err
+	}
+
+	if len(operatorGroupList.Items) == 1 && operatorGroupList.Items[0].UpgradeStrategy() == operatorsv1.UnsafeFailForwardUpgradeStrategy {
+		csvs, err := r.csvLister.ClusterServiceVersions(namespace).List(labels.Everything())
+		if err != nil {
+			return predicates, err
+		}
+
+		replacementMapping := map[string]*v1alpha1.ClusterServiceVersion{}
+		for _, csv := range csvs {
+			if csv.Spec.Replaces != "" {
+				replacementMapping[csv.Spec.Replaces] = csv
+			}
+		}
+
+		for i := range csvs {
+			if !csvs[i].IsCopied() {
+				if csvs[i].Status.Phase == v1alpha1.CSVPhaseReplacing && replacementChainEndsInFailure(csvs[i], replacementMapping) {
+					predicates = append(predicates, cache.Not(cache.CSVNamePredicate(csvs[i].GetName())))
```

Comment:
I see what you are trying to do here but I'm a bit iffy with using `Not` predicate given it bit us in the butt once with Eric's work. Frankly, I wonder if we simply exclude all `Replacing` CSV from resolution. Would that be better? I personally want to have Ben's opinion here but he is unavailable.
I think it will be OK with this solution if we have a solid e2e test case to ensure this will work.